thriftclient/client_pool.go

```diff
@@ -0,0 +1,211 @@
+package thriftclient
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	"github.com/apache/thrift/lib/go/thrift"
+	"github.com/go-kit/kit/metrics"
+
+	"github.com/reddit/baseplate.go/clientpool"
+	"github.com/reddit/baseplate.go/log"
+	"github.com/reddit/baseplate.go/metricsbp"
+)
+
+// ClientPoolConfig is the configuration struct for creating a new ClientPool.
+type ClientPoolConfig struct {
+	// ServiceSlug is a short identifier for the thrift service you are creating
+	// clients for.  The preferred convention is to take the service's name,
+	// remove the 'Service' prefix, if present, and convert from camel case to
+	// all lower case, hyphen separated.
+	//
+	// Example:
+	//		AuthenticationService -> authentication
+	//		ImageUploadService -> image-upload
+	ServiceSlug string
+
+	// Addr is the address of a thrift service.  Addr must be in the format
+	// ${host}:${port}
+	Addr string
+
+	// MinConnections is the minimum number of thrift connections the client
+	// pool should maintain.
+	MinConnections int
+
+	// MinConnections is the maximum number of thrift connections the client
+	// pool can maintain.
+	MaxConnections int
+
+	// SocketTimeout is the timeout on the underling thrift.TSocket.
+	SocketTimeout time.Duration
+
+	// ReportPoolStats signals to the ClientPool that it should report
+	// statistics on the underlying clientpool.Pool in a background
+	// goroutine.  If this is set to false, the reporting goroutine will
+	// not be started and it will not report pool stats.
+	//
+	// It reports the number of active clients to a gauge named
+	// "${ServiceSlug}.pool-active-connections".
+	//
+	// The reporting goroutine is cancelled when the global metrics client
+	// context is Done.
+	ReportPoolStats bool
+
+	// PoolGaugeInterval indicates how often we should update the active
+	// connections gauge when collecting pool stats.
+	PoolGaugeInterval time.Duration
+}
+
+// Client is a client object that implements both the clientpool.Client and
+// thrift.TCLient interfaces.
+//
+// This allows it to be managed by a clientpool.Pool and be passed to a thrift
+// client as the base thrift.TClient.
+type Client interface {
+	clientpool.Client
+	thrift.TClient
+}
+
+// ClientPool defines an object that can be used to manage a pool of
+// Client objects.
+type ClientPool interface {
+	// GetClient returns a Client from the pool or creates a new one if
+	// needed.
+	GetClient() (Client, error)
+
+	// ReleaseClient returns the given client to the pool.
+	ReleaseClient(Client)
+}
+
+// AddressGenerator defines a function that returns the address of a thrift
+// service.
+type AddressGenerator func() string
```

Comment:
I think this generator needs to return `(string, error)` instead. For `SingleAddressGenerator` it can be that it always return nil error.